lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp

```diff
@@ -16,103 +19,331 @@ using namespace mlir;
+      b.mapping.map(forOp.getResult(oldIdx), newForOp.getResult(newIdx));
+    }
+
+    b.setInsertionPointToStart(newForOp.getBody());
   }
 
-  // Delete everything else.
-  eraseOtherPartitions(loop, schedule, partition);
+  cloneOpsInBlock(forOp.getBody(), builders, schedule);
 
-  return std::move(block);
+  for (auto newForOp : newForOps) {
+    WarpSchedule::eraseFrom(newForOp);
+  }
 }
 
-//===----------------------------------------------------------------------===//
-// partitionLoop
-//===----------------------------------------------------------------------===//
+void mapRange(ValueRange fromRange, ValueRange toRange, IRMapping &mapping) {
+  for (auto [from, to] : llvm::zip(fromRange, toRange)) {
+    mapping.map(from, to);
+  }
+}
+
+SmallVector<size_t>
+getPartitionIndicesToCloneInto(const Partition *partition,
+                               const WarpSchedule &schedule) {
+  SmallVector<size_t> partitionIndices;
+
+  if (partition == schedule.getRootPartition()) {
+    for (size_t i = 0; i < schedule.getNumPartitions(); ++i) {
+      partitionIndices.push_back(i);
+    }
+  } else {
+    partitionIndices.push_back(partition->getIndex());
+  }
+
+  return partitionIndices;
+}
+
+void cloneIfOp(scf::IfOp ifOp, SmallVector<WarpGroupBuilder> &builders,
+               const WarpSchedule &schedule) {
+  auto partition = getPartition(ifOp, schedule);
+  assert(partition);
+  auto partitionIndices = getPartitionIndicesToCloneInto(partition, schedule);
+
+  SmallVector<scf::IfOp> newIfOps;
+  for (size_t idx : partitionIndices) {
+    auto &b = builders[idx];
+    auto cond = b.mapping.lookupOrDefault(ifOp.getCondition());
+    auto newIfOp = b.create<scf::IfOp>(ifOp.getLoc(), ifOp.getResultTypes(),
+                                       cond, ifOp.elseBlock() ? true : false);
+    newIfOp->setAttrs(ifOp->getAttrs());
+    newIfOps.push_back(newIfOp);
+
+    mapRange(ifOp.getResults(), newIfOp.getResults(), b.mapping);
+    mapRange(ifOp.thenBlock()->getArguments(),
+             newIfOp.thenBlock()->getArguments(), b.mapping);
+
+    if (ifOp.elseBlock()) {
+      mapRange(ifOp.elseBlock()->getArguments(),
+               newIfOp.elseBlock()->getArguments(), b.mapping);
+    }
+
+    b.setInsertionPointToStart(newIfOp.thenBlock());
+  }
+
+  cloneOpsInBlock(ifOp.thenBlock(), builders, schedule);
+
+  if (auto elseBlock = ifOp.elseBlock()) {
+    for (auto [builder, newIfOp] : llvm::zip(builders, newIfOps)) {
+      builder.setInsertionPointToStart(newIfOp.elseBlock());
+    }
+    cloneOpsInBlock(elseBlock, builders, schedule);
+  }
+
+  for (auto [idx, newIfOp] : llvm::zip(partitionIndices, newIfOps)) {
+    builders[idx].setInsertionPointAfter(newIfOp);
+  }
+}
+
+void cloneReduceOp(triton::ReduceOp reduceOp,
+                   SmallVector<WarpGroupBuilder> &builders,
+                   const WarpSchedule &schedule) {
+  auto partition = getPartition(reduceOp, schedule);
+  assert(partition);
```

Comment:
at this point I think it would make sense for the assert to be folded inside the `getPartition` function